File: framework/languages/node/e2e/AutomaticTurnDispatch/Server/Play/play-host-factory.ts
import fs from 'node:fs';
import { Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { ZLinkMessageFlowLogMode, ZLinkUserSpotExecutionMode, type ZLinkFrameworkRuntime } from '@zlink-systems/framework';
import { ZLinkRedisLocationStore } from '@zlink-systems/framework-locations-redis';
import { ZLinkHttpClientModule, ZLinkModule, ZLINK_FRAMEWORK_RUNTIME, zlinkFramework } from '@zlink-systems/nestjs';
import { AutomaticTurnDispatchNames } from '../../Shared/messages';
import { EvidenceStore } from './Support/evidence-store';
import { closeHttpServer, startHttpServer } from './Support/http-server';
import { createAutomaticTurnConfiguration } from '../../configuration';
import { PLAY_OPTIONS, validatePlayOptions } from './Configuration/play-options';
import type { PlayOptions } from './Configuration/play-options';
import { HoldCommandHandler, ProbeCommandHandler, ProbeRequestHandler, WorkerAwaitCommandHandler, AwaitCommandHandler, AwaitRequestHandler } from './Handlers/basic-spot-handlers';
import {
  CounterAwaitHandler,
  CounterReadHandler,
  CounterResetHandler,
  CpuWorkerAwaitHandler,
  HttpAwaitHandler,
  IoWorkerBatchHandler,
  SelfCycleHandler,
  SelfSendHandler
} from './Handlers/execution-turn-handlers';
import {
  BindAwaitActorsControlHandler,
  EnsureSpotControlHandler,
  YIELD_PLAY_NODE_RID,
  AwaitEvidenceControlHandler,
  AwaitEvidenceWaitControlHandler
} from './Handlers/control-handlers';
import { AwaitCancelCommandHandler, AwaitTimeoutCommandHandler } from './Handlers/failure-spot-handlers';
import { RemoteSpotAwaitCommandHandler, RemoteSpotAwaitHandler } from './Handlers/remote-spot-handlers';
import { TimerStartCommandHandler, TimerStopCommandHandler, AwaitTimerHandler } from './Handlers/timer-spot-handlers';
import {
  EntryActorFastHandler,
  EntryActorFastSendHandler,
  EntryActorJoinAwaitHandler,
  EntryActorPushAwaitHandler,
  EntryActorAwaitHandler,
  SpotActorFastHandler,
  SpotActorFastSendHandler,
  SpotActorDeferredJoinFailureHandler,
  SpotActorJoinAwaitHandler,
  SpotActorPushAwaitHandler,
  SpotActorAwaitHandler,
  AwaitActorFactory,
  AwaitEntrySpot
} from './Spots/await-actors';
import { AwaitProbeSpot } from './Spots/await-probe-spot';
import { PerActorAwaitProbeSpot } from './Spots/per-actor-await-probe-spot';

export async function startPlayHost(): Promise<void> {
  let stopping = false;
  const configured = createAutomaticTurnConfiguration(PLAY_OPTIONS, validatePlayOptions);
  const configuration = configured.module;
  const frameworkModule = ZLinkModule.forRootFactory({
    imports: [configuration],
    inject: [PLAY_OPTIONS],
    useFactory: (value: unknown) => {
      const options = value as PlayOptions;
      fs.mkdirSync(options.logDir, { recursive: true });
      const locationStore = new ZLinkRedisLocationStore({
        url: `redis://${options.redisEndpoint}`,
        keyPrefix: options.redisKeyPrefix
      });
      const builder = zlinkFramework();
      builder
        .configureDispatch()
          .messageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
          .traceLogFile(`${options.logDir}/${options.rid}-flow.log`)
          .traceLabel(options.rid);
      builder.addLocationStore(locationStore);
      builder.addRouteMesh(AutomaticTurnDispatchNames.controlChannel)
        .listen(options.controlEndpoint)
        .routingId(options.rid)
        .addRequestHandler('EnsureSpotReq', EnsureSpotControlHandler)
        .addRequestHandler('BindAwaitActorsReq', BindAwaitActorsControlHandler)
        .addRequestHandler('AwaitEvidenceReq', AwaitEvidenceControlHandler)
        .addRequestHandler('AwaitEvidenceWaitReq', AwaitEvidenceWaitControlHandler)
        .channel(AutomaticTurnDispatchNames.controlChannel).server();
      const delayMesh = builder.addRouteMesh(AutomaticTurnDispatchNames.delayChannel);
      delayMesh.peerConnections().connect(options.delayEndpoint);
      delayMesh.channel(AutomaticTurnDispatchNames.delayChannel).client();
      const spotRoute = builder.addRouteMesh(AutomaticTurnDispatchNames.spotRouteChannel)
        .listen(options.spotRouteEndpoint)
        .routingId(options.rid);
      spotRoute.channel(AutomaticTurnDispatchNames.spotRouteChannel).server();
      for (const endpoint of options.peerSpotRouteEndpoints) {
        spotRoute.peerConnections().connect(endpoint);
      }
      const spotMesh = builder.addRouteMesh(AutomaticTurnDispatchNames.spotChannel)
        .routingId(options.rid)
        .listen(options.spotRouterEndpoint);
      const objectServer = spotMesh.objects().server();
      objectServer.addEntrySpot(AwaitEntrySpot);
      objectServer.addActorFactory(
        AutomaticTurnDispatchNames.actorType,
        AwaitActorFactory,
        (factory) => factory.disableRelocation()
      );
      objectServer.addSpotFactory(
        AwaitProbeSpot.name,
        AwaitProbeSpot,
        (factory) => factory.disableRelocation()
      );
      objectServer.addSpotFactory(
        AutomaticTurnDispatchNames.perActorSpotType,
        PerActorAwaitProbeSpot,
        (factory) => {
          factory.executionMode(ZLinkUserSpotExecutionMode.PerActor);
          factory.recreateOnRelocation();
        }
      );
      spotMesh.channel(AutomaticTurnDispatchNames.spotChannel).server();
      for (const peer of options.spotRouterPeers) spotMesh.peerConnections().connect(peer.rid, peer.endpoint);
      return builder.build();
    }
  });
  const httpClientModule = ZLinkHttpClientModule.forRoot({
    imports: [frameworkModule],
    clients: [{ name: 'external-api', baseUrl: configured.options.externalApiUrl }]
  });

  class PlayModule {}
  Module({
    imports: [
      configuration,
      frameworkModule,
      httpClientModule
    ],
    providers: [
      {
        provide: EvidenceStore,
        inject: [PLAY_OPTIONS],
        useFactory: (options: PlayOptions) => new EvidenceStore(options.rid, options.evidenceFile)
      },
      {
        provide: YIELD_PLAY_NODE_RID,
        inject: [PLAY_OPTIONS],
        useFactory: (options: PlayOptions) => options.rid
      },
      EnsureSpotControlHandler,
      BindAwaitActorsControlHandler,
      AwaitEvidenceControlHandler,
      AwaitEvidenceWaitControlHandler,
      HoldCommandHandler,
      AwaitCommandHandler,
      AwaitRequestHandler,
      ProbeRequestHandler,
      CounterResetHandler,
      CounterAwaitHandler,
      CounterReadHandler,
      HttpAwaitHandler,
      IoWorkerBatchHandler,
      CpuWorkerAwaitHandler,
      SelfCycleHandler,
      SelfSendHandler,
      WorkerAwaitCommandHandler,
      AwaitTimeoutCommandHandler,
      AwaitCancelCommandHandler,
      ProbeCommandHandler,
      RemoteSpotAwaitHandler,
      RemoteSpotAwaitCommandHandler,
      TimerStartCommandHandler,
      TimerStopCommandHandler,
      AwaitTimerHandler,
      AwaitActorFactory,
      AwaitEntrySpot,
      EntryActorAwaitHandler,
      EntryActorFastHandler,
      EntryActorFastSendHandler,
      EntryActorJoinAwaitHandler,
      EntryActorPushAwaitHandler,
      SpotActorAwaitHandler,
      SpotActorFastHandler,
      SpotActorFastSendHandler,
      SpotActorDeferredJoinFailureHandler,
      SpotActorJoinAwaitHandler,
      SpotActorPushAwaitHandler,
      AwaitProbeSpot,
      PerActorAwaitProbeSpot
    ]
  })(PlayModule);

  const app = await NestFactory.createApplicationContext(PlayModule, { logger: false, abortOnError: false });
  const options = app.get<PlayOptions>(PLAY_OPTIONS);
  const evidence = app.get(EvidenceStore);
  const frameworkRuntime = app.get<ZLinkFrameworkRuntime>(ZLINK_FRAMEWORK_RUNTIME, { strict: false });
  let shutdownOperation: Promise<unknown> | undefined;
  let shutdownError: unknown;
  const server = await startHttpServer(options.httpUrl, [
    {
      method: 'GET',
      path: '/health',
      handle: () => ({
        status: frameworkRuntime.status.acceptingWork ? 'ready' : 'draining',
        role: 'play',
        rid: options.rid
      })
    },
    { method: 'GET', path: '/evidence', handle: () => evidence.snapshot() },
    {
      method: 'GET',
      path: '/status',
      handle: () => ({
        state: frameworkRuntime.status.state,
        acceptingWork: frameworkRuntime.status.acceptingWork,
        sequence: frameworkRuntime.status.sequence.toString()
      })
    },
    {
      method: 'POST',
      path: '/shutdown',
      handle: () => {
        if (shutdownOperation === undefined) {
          shutdownOperation = frameworkRuntime.shutdown({ deadlineMs: 30_000 });
          void shutdownOperation.then(
            () => { stopping = true; },
            (error: unknown) => { shutdownError = error; stopping = true; }
          );
        }
        return {
          status: 'draining',
          acceptingWork: frameworkRuntime.status.acceptingWork,
          state: frameworkRuntime.status.state
        };
      }
    }
  ]);
  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  if (shutdownError !== undefined) throw shutdownError;
  await app.close();
}

File: framework/languages/dotnet/e2e/AutomaticTurnDispatch/Server/Play/PlayHostFactory.cs
using Microsoft.Extensions.Configuration;

using Systems.Zlink;
using AutomaticTurnDispatch.Server.Play.Handlers;
using AutomaticTurnDispatch.Server.Play.Spots;
using AutomaticTurnDispatch.Shared;
using Zlink.Framework.AspNetCore;
using Zlink.Framework.Locations.Redis;
using Zlink.Framework.Contracts.Configuration;
using Zlink.Framework.Contracts.Dispatch;
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.Contracts.Channels;
using Zlink.Framework.Contracts.Errors;
using Zlink.Framework.Contracts.Spots;
using Zlink.Framework.E2E.Diagnostics;

namespace AutomaticTurnDispatch.Server.Play;

internal static class PlayHostFactory
{
    public static WebApplication Create(string[] args)
    {
        var options = PlayOptions.Parse(args);
        Directory.CreateDirectory(options.LogDir);

        var builder = WebApplication.CreateBuilder(args);
        builder.Configuration.Sources.Clear();
        builder.Configuration.AddInMemoryCollection();
        builder.Logging.ClearProviders();
        builder.Logging.AddSimpleConsole(console =>
        {
            console.SingleLine = true;
            console.TimestampFormat = "HH:mm:ss.fff ";
        });
        builder.WebHost.UseUrls(options.HttpUrl);
        builder.Services.AddSingleton(new EvidenceStore(options.Rid, options.EvidenceFile));
        builder.Services.AddSingleton(new NodeOptions(options.Rid));
        builder.Services.AddSingleton(new E2eMessageFlowListener(
            Path.Combine(options.LogDir, $"{options.Rid}-flow.log"),
            options.Rid));
        builder.Services.AddZLinkHttpClient("external-api", http => http
            .BaseUrl(options.ExternalApiBaseUrl)
            .Timeout(TimeSpan.FromSeconds(5)));
        builder.Services.AddZLinkFramework(framework =>
        {
            //  This E2E host is not started inside a memory-limited
            //  container. Supply a deterministic finite limit so the
            //  default Auto HWM contract does not depend on the host.
            framework.ConfigureInboundDispatch().ProcessMemoryLimitBytes =
                1UL * 1024 * 1024 * 1024;
            // Actor Join callbacks intentionally hold admission for 350 ms.
            // Leave deterministic time for transport and Store work.
            framework.DefaultRequestTimeout = TimeSpan.FromSeconds(2);
            framework.AddLocationStore(new ZLinkRedisLocationStore(redis => { redis.ConnectionString = options.RedisEndpoint; redis.KeyPrefix = options.RedisKeyPrefix; }));
            framework.AddRelocationStore(new ZLinkRedisRelocationStore(redis => { redis.ConnectionString = options.RedisEndpoint; redis.KeyPrefix = $"{options.RedisKeyPrefix}:relocation"; }));
            framework.AddHandlersFromAssemblyOf(typeof(Program));
            framework.ConfigureDispatch().Diagnostics
                .SetLevel(ZLinkDiagnosticsLevel.Normal);
            var controlMesh = framework.AddRouteMesh(AutomaticTurnDispatchNames.ControlChannel)
                .Listen(options.ControlEndpoint)
                .SetRoutingIdPrefix(options.Rid);
            controlMesh.Channel(AutomaticTurnDispatchNames.ControlChannel).Server();
            controlMesh
                .AddRouteRequestHandler<BindAwaitActorsControlHandler, BindAwaitActorsReq, BindAwaitActorsRes>(
                    "BindAwaitActorsReq")
                .AddRouteRequestHandler<EnsureSpotControlHandler, EnsureSpotReq, EnsureSpotRes>("EnsureSpotReq")
                .AddRouteRequestHandler<AwaitEvidenceControlHandler, AwaitEvidenceReq, AwaitEvidenceRes>(
                    "AwaitEvidenceReq")
                .AddRouteRequestHandler<AwaitEvidenceWaitControlHandler, AwaitEvidenceWaitReq, AwaitEvidenceRes>(
                    "AwaitEvidenceWaitReq");
            var delayMesh = framework.AddRouteMesh(AutomaticTurnDispatchNames.DelayChannel)
                .Listen("tcp://127.0.0.1:0")
                .SetRoutingId(RoutingId.From(options.Rid));
            delayMesh.Channel(AutomaticTurnDispatchNames.DelayChannel).Client();
            delayMesh.PeerConnections.Connect(options.DelayEndpoint);
            var spotRouteMesh = framework.AddRouteMesh(AutomaticTurnDispatchNames.SpotRouteChannel)
                .Listen(options.SpotRouteEndpoint)
                .SetRoutingIdPrefix(options.Rid);
            spotRouteMesh.Channel(AutomaticTurnDispatchNames.SpotRouteChannel).Server();
            var mesh24 = framework.AddRouteMesh(AutomaticTurnDispatchNames.SpotChannel)
                .Listen(options.SpotRouterEndpoint)
                .SetRoutingIdPrefix(options.Rid)
                .SetPlacementWeight(options.PlacementWeight);
            mesh24.Objects().Server()
                .AddEntrySpot<AwaitEntrySpot>()
                .AddActorFactory<AwaitActor, AwaitActorFactory>(
                    AutomaticTurnDispatchNames.ActorType,
                    factory => factory.RecreateOnRelocation())
                .AddSpotFactory<AwaitProbeSpot>(
                    AutomaticTurnDispatchNames.SpotType,
                    factory => factory.DisableRelocation())
                .AddSpotFactory<PerActorAwaitSpot>(
                    AutomaticTurnDispatchNames.PerActorSpotType,
                    factory => factory
                        .ExecutionMode(ZLinkUserSpotExecutionMode.PerActor)
                        .RecreateOnRelocation());
            mesh24.Channel(AutomaticTurnDispatchNames.SpotChannel).Server();
        });

        var app = builder.Build();
        app.MapGet("/health", () => Results.Ok(new { status = "ready", role = "play", options.Rid }));
        app.MapGet("/topology/ready", async (
            string meshName,
            string rid,
            IZLinkRouteMeshRuntime runtime,
            IZLinkRouteClient routes,
            CancellationToken cancellationToken) =>
        {
            if (string.Equals(
                    meshName,
                    AutomaticTurnDispatchNames.DelayChannel,
                    StringComparison.Ordinal))
            {
                try
                {
                    _ = await routes.RequestToChannel(
                            meshName,
                            new DelayReq($"readiness-{rid}", 0, "readiness"))
                        .Timeout(TimeSpan.FromMilliseconds(500))
                        .Async<DelayRes>(cancellationToken);
                    return Results.Ok(new { ready = true });
                }
                catch (Exception error) when (
                    error is ZLinkFrameworkException
                        or TimeoutException
                        or OperationCanceledException)
                {
                    return Results.Ok(new { ready = false });
                }
            }

            var snapshot = runtime.GetStatus(meshName);
            var ready = snapshot.Peers.Any(peer =>
                peer.State == ZLinkPeerState.Ready
                && (string.Equals(
                        peer.NodeRid.ToString(),
                        rid,
                        StringComparison.Ordinal)
                    || peer.NodeRid.ToString().StartsWith(
                        $"{rid}-",
                        StringComparison.Ordinal)));
            return Results.Ok(new { ready });
        });
        app.MapPost("/placement-weight", async (
            PlacementWeightReq request,
            IZLinkRouteMeshRuntimeOptions runtimeOptions,
            IZLinkSpotManager spots,
            NodeOptions node,
            CancellationToken cancellationToken) =>
        {
            var placement = runtimeOptions.Mesh(AutomaticTurnDispatchNames.SpotChannel);
            placement.PlacementWeight = request.Weight;
            if (request.Weight > 0 && request.VerifyLocal)
            {
                var deadline = DateTimeOffset.UtcNow.AddSeconds(10);
                var consecutiveLocal = 0;
                while (DateTimeOffset.UtcNow < deadline)
                {
                    ZLinkSpotCreateResult probe;
                    try
                    {
                        probe = await spots.GetOrCreate(
                                $"placement-probe-{Guid.NewGuid():N}",
                                AutomaticTurnDispatchNames.SpotType)
                            .Timeout(TimeSpan.FromSeconds(2))
                            .Async(cancellationToken);
                    }
                    catch (ZLinkFrameworkException error)
                        when (error.Kind is
                            ZLinkFrameworkErrorKind.Unavailable
                            or ZLinkFrameworkErrorKind.CapacityExceeded
                            or ZLinkFrameworkErrorKind.DeadlineExceeded)
                    {
                        await Task.Delay(25, cancellationToken);
                        continue;
                    }
                    var local = probe.Spot.NodeRid.ToString().StartsWith(
                        $"{node.Rid}-",
                        StringComparison.Ordinal);
                    await spots.CloseAsync(probe.Spot, cancellationToken);
                    consecutiveLocal = local ? consecutiveLocal + 1 : 0;
                    if (consecutiveLocal >= 4)
                        return Results.Ok(new PlacementWeightRes(placement.PlacementWeight));
                    await Task.Delay(25, cancellationToken);
                }

                return Results.StatusCode(StatusCodes.Status503ServiceUnavailable);
            }

            return Results.Ok(new PlacementWeightRes(placement.PlacementWeight));
        });
        app.MapGet("/evidence", (EvidenceStore evidence) => Results.Ok(evidence.Snapshot()));
        return app;
    }
}
